magtogoek/tools.py

```diff
@@ -180,4 +181,307 @@ def rotate_2d_vector(
+
+    return result
+
+
+def _new_flags_bin_regrid(flags: xr.DataArray,
+                          bin_depths: np.ndarray,
+                          dim: str = 'depth',
+                          axis: int = 0) -> xr.DataArray:
+    """
+    Quality flag transfer function for bin average regridding.
+
+    Parameters
+    ----------
+    flags :
+        Flags prepared for regridding (5, 8, 9).
+    bin_depths :
+        Center of the regridding averaging bins (meters).
+    dim :
+        Name of the dimension along which the binning is done.
+    axis :
+        Index of the dimension along which the binning is done.
+
+    Returns
+    -------
+    new_flags :
+        Quality flags for the regridded data.
+
+    """
+    bin_edges = _bin_centers_to_edges(bin_depths)
+
+    # Determine if bins contained 5 or 8 flags
+    grouped = flags.groupby_bins(dim, bin_edges, labels=bin_depths)
+    contains_8 = grouped.map(_isin_along_axis, args=(8, axis,))
+    contains_5 = grouped.map(_isin_along_axis, args=(5, axis,))
+
+    # Make new flag array
+    new_flags = xr.where(contains_5, 5, 9)  # contains 5 or 8 -> 5, not -> 9
+    new_flags = xr.where(contains_8, 8, new_flags)  # contains 8 -> 8
+
+    return new_flags
+
+
+def _new_flags_interp_regrid(dataset: xr.Dataset, variable: str) -> xr.DataArray:
+    """
+    Quality flag transfer function for interpolation regridding.
+
+    Parameters
+    ----------
+    flags :
+        Dataset being regridded.
+    variable :
+        Name of the variable for which flags are being transfered.
+
+    Returns
+    -------
+    new_flags :
+        Quality flags for the regridded data variable.
+
+    """
+    condition_null = dataset[f"{variable}"].isnull()
+    condition_good = (dataset[f"{variable}_QC"] == 8) & ~condition_null
+    condition_changed = (dataset[f"{variable}_QC"] == 5) & condition_null
+    new_flags = dataset[f"{variable}_QC"].copy()
+    new_flags.loc[:] = 9
+    new_flags = xr.where(condition_changed, 5, new_flags)
+    new_flags = xr.where(condition_good, 8, new_flags)
+
+    return new_flags
+
+
+def _prepare_flags_for_regrid(flags: xr.DataArray) -> xr.DataArray:
+    """
+    Format quality control flags for input to regridding transfer schemes.
+
+    Parameters
+    ----------
+    flags :
+        Quality control flags [0, 1, 2, 3, 4, 5, 9].
+
+    Returns
+    -------
+    new_flags :
+        Quality control flags [5, 8, 9].
+
+    Notes
+    -----
+    The rationale for this transformation is that after interpolation or
+    averaging missing values should remain flagged as `missing`, good values
+    should be flagged as `interpolated` and questionable values should be
+    flagged as `modified`. All flag values are therefore funneled into
+    these three types before regridding to simplify the transfer algorithms.
+    """
+    new_flags = xr.where(flags < 9, 5, 9)
+    new_flags = xr.where(flags < 3, 8, new_flags)
```

Comment:
why not just work in numpy array

```suggestion
    new_flags = 9*np.ones(flags.shape)
    new_flags[flags < 9] = 5
    new_flags[flags < 3] = 8
```

Reply:
This segment was modified according to offline suggestions.